README.md
# Iker

Wrapper tool around ike-scan to help analyse the security of an IPsec based VPN endpoint.

This is a rewrite of the original tool initially developer by @julgoor

Released under th GPL v3 license <http://www.gnu.org/licenses/gpl-3.0.html>

### Funcionalities

* IPSec discovering
* IKE v2 support detection
* Vendor IDs (VID) extraction
* IPSec implementation guessing (backoff)
* IPSec main mode transforms listing
* IPSec aggressive mode transforms listing
* IPSec aggressive mode client/group IDs listing
* XML and text output

### Usage

```bash
$ python iker.py -i ips.txt -o iker_output.txt -x iker_output.xml -v
```

### Authors

* Borja <borja [at] libcrack [dot] so>
* Julio Gomez  <julgor [at] gmail [dot] com>
* Pablo Cat  <xkill [at] locolandia [dot] net>


CHANGES
* [1;34md0f45c2[m - [1;36mTue, 8 Dec 2015 21:34:24 +0100[m [1;32m(2 minutes ago)[m[1;33m (HEAD -> development, origin/development)[m
|           [37mImplemented cli function[m [2;37m- Borja R[m
* [1;34m67447ac[m - [1;36mSat, 26 Sep 2015 13:09:30 +0200[m [1;32m(2 months ago)[m[1;33m[m
|           [37mchg: TODO.md - Added 4 new bullet points to 2do list[m [2;37m- Borja[m
* [1;34m5b54b26[m - [1;36mSat, 26 Sep 2015 13:08:08 +0200[m [1;32m(2 months ago)[m[1;33m[m
|           [37mnew: CLI added: cli.py loads iker.py:main()[m [2;37m- Borja[m
* [1;34m437a1a2[m - [1;36mSat, 26 Sep 2015 13:05:27 +0200[m [1;32m(2 months ago)[m[1;33m[m
|           [37mchg: Function check_privileges supressed, funcionality implemented inside iker.py main() method[m [2;37m- Borja[m
* [1;34mde88bbb[m - [1;36mSat, 26 Sep 2015 12:47:15 +0200[m [1;32m(2 months ago)[m[1;33m[m
|           [37mnew: Added git submodule ike-scan[m [2;37m- Borja[m
* [1;34mbf62861[m - [1;36mSat, 26 Sep 2015 12:22:55 +0200[m [1;32m(2 months ago)[m[1;33m[m
|           [37mchg: global VERBOSE && -v/--verbose arg option deleted new: added -l/--loglevel argument option chg: References to "print_message" substituyed by logger.<loglevel>(msg)[m [2;37m- Borja[m
* [1;34m0dd8ecf[m - [1;36mSat, 26 Sep 2015 12:18:24 +0200[m [1;32m(2 months ago)[m[1;33m[m
|           [37mnew: Added logging support via custom logger.py module[m [2;37m- Borja[m
* [1;34me0df4f6[m - [1;36mSat, 26 Sep 2015 10:48:16 +0200[m [1;32m(2 months ago)[m[1;33m[m
|           [37mnew: Creation of "iker" setuptools package[m [2;37m- Borja[m
* [1;34m2ef35fe[m - [1;36mSat, 26 Sep 2015 10:47:20 +0200[m [1;32m(2 months ago)[m[1;33m[m
|           [37mnew: Creation of "iker" python package[m [2;37m- Borja[m
* [1;34m2f59e5b[m - [1;36mSat, 26 Sep 2015 10:42:46 +0200[m [1;32m(2 months ago)[m[1;33m[m
|           [37mchg: TODO.md minor modification[m [2;37m- Borja[m
* [1;34m635de0b[m - [1;36mSat, 26 Sep 2015 10:41:34 +0200[m [1;32m(2 months ago)[m[1;33m[m
|           [37mchg: Rename virtualenv "venv" to "env"[m [2;37m- Borja[m
* [1;34mcd1a451[m - [1;36mSat, 26 Sep 2015 10:36:34 +0200[m [1;32m(2 months ago)[m[1;33m[m
|           [37mnew: Added auto PEP8 script[m [2;37m- Borja[m
* [1;34m4b9413e[m - [1;36mSat, 26 Sep 2015 10:36:15 +0200[m [1;32m(2 months ago)[m[1;33m[m
|           [37mfix: Rename TODO -> TODO.md & iker.py -> iker/iker.py[m [2;37m- Borja[m
* [1;34m48e8312[m - [1;36mSat, 26 Sep 2015 08:31:26 +0200[m [1;32m(2 months ago)[m[1;33m (origin/master, origin/HEAD, master)[m
|           [37mnew: Moved description from iker.py to README.md[m [2;37m- Borja[m
* [1;34m88dabbb[m - [1;36mSat, 26 Sep 2015 08:30:36 +0200[m [1;32m(2 months ago)[m[1;33m[m
|           [37mnew: Added TODO[m [2;37m- Borja[m
* [1;34mca758db[m - [1;36mSat, 26 Sep 2015 08:04:54 +0200[m [1;32m(2 months ago)[m[1;33m[m
|           [37mnew: Changelog support[m [2;37m- Borja[m
* [1;34mb020fc3[m - [1;36mFri, 18 Sep 2015 09:10:23 +0200[m [1;32m(3 months ago)[m[1;33m[m
|           [37mPermission +x to iker.py[m [2;37m- Borja[m
* [1;34m51c695f[m - [1;36mFri, 18 Sep 2015 09:08:24 +0200[m [1;32m(3 months ago)[m[1;33m[m
|           [37mMinor modification to README.md[m [2;37m- Borja[m
* [1;34md8ae1d5[m - [1;36mFri, 18 Sep 2015 09:07:02 +0200[m [1;32m(3 months ago)[m[1;33m[m
|           [37mCode refactorized to meet PEP8 reqs[m [2;37m- Borja[m
* [1;34mfdb7047[m - [1;36mFri, 18 Sep 2015 08:55:27 +0200[m [1;32m(3 months ago)[m[1;33m[m
|           [37mCode refactorized to meet PEP8 reqs[m [2;37m- Borja[m
* [1;34mbc83e5b[m - [1;36mFri, 18 Sep 2015 08:29:10 +0200[m [1;32m(3 months ago)[m[1;33m[m
|           [37mAvoid use of <tabs>[m [2;37m- Borja[m
* [1;34mda138ea[m - [1;36mFri, 18 Sep 2015 08:27:19 +0200[m [1;32m(3 months ago)[m[1;33m[m
|           [37mRefactorisation of "delay"[m [2;37m- Borja[m
* [1;34md2e9dac[m - [1;36mFri, 18 Sep 2015 08:23:11 +0200[m [1;32m(3 months ago)[m[1;33m[m
|           [37mAdapted function names to PEP[m [2;37m- Borja[m
* [1;34m5a2e911[m - [1;36mFri, 18 Sep 2015 08:16:12 +0200[m [1;32m(3 months ago)[m[1;33m[m
|           [37mRefactoring module call (sys,os) in a pythonic way[m [2;37m- Borja[m
* [1;34mc0b1d5e[m - [1;36mWed, 8 Jul 2015 18:47:13 +0200[m [1;32m(5 months ago)[m[1;33m[m
|           [37mInitial commit[m [2;37m- Borja[m
* [1;34m1042418[m - [1;36mWed, 8 Jul 2015 18:46:00 +0200[m [1;32m(5 months ago)[m[1;33m[m
            [37mInitial commit[m [2;37m- Fantastico[m


iker/iker.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

import subprocess
import argparse
import time
import sys
import os
import re

from . import color
from . import ike_scan_wrapper
from .logger import Logger

logger = Logger.logger


def banner():
    """
    Prints a banner message.
    """
    print("ike-scan wrapper for testing IPsec-based VPNs.")


def get_arguments():
    """
    Parses command line options
        @returns the arguments received and a list of targets.
    """
    targets = []
    parser = argparse.ArgumentParser()

    parser.add_argument(
        "target",
        type=str,
        nargs='?',
        help="The IP address or the network (CIDR notation) to scan.")
    parser.add_argument(
        "-l",
        "--loglevel",
        action="store",
        type=str,
        default='info',
        help="logging verbose level")
    parser.add_argument(
        "-d",
        "--delay",
        type=int,
        help="Delay between requests (in milliseconds). Default: 0 (No delay).")
    parser.add_argument(
        "-i",
        "--input",
        type=str,
        default=None,
        help="An input file with an IP address/network per line.")
    parser.add_argument(
        "-o",
        "--output",
        type=str,
        help="An output file to store the results.")
    parser.add_argument(
        "-x",
        "--xml",
        type=str,
        help="An output file to store the results in XML format. Default: output.xml")
    parser.add_argument(
        "--encalgs",
        type=str,
        default="1 5 7/128 7/192 7/256",
        help="The encryption algorithms to check. Default: DES, 3DES, AES/128, AES/192 and AES/256. "
        "Example: --encalgs=\"1 5 7/128 7/192 7/256\"")
    parser.add_argument(
        "--hashalgs",
        type=str,
        default="1 2",
        help="The hash algorithms to check. Default: MD5 and SHA1. Example: --hashalgs=\"1 2\"")
    parser.add_argument(
        "--authmethods",
        type=str,
        default="1 3 64221 65001",
        help="The authorization methods to check. Default: Pre-Shared Key, RSA Signatures, Hybrid Mode and XAUTH. "
        "Example: --authmethods=\"1 3 64221 65001\"")
    parser.add_argument(
        "--dhgroups",
        type=str,
        default="1 2 5",
        help="The Diffie-Hellman groups to check. Default: MODP 768, MODP 1024 and MODP 1536. "
        "Example: --dhgroups=\"1 2 5\"")
    parser.add_argument(
        "--fullalgs",
        action="store_true",
        help="Equivalent to: "
        "--encalgs=\"1 2 3 4 5 6 7/128 7/192 7/256 8\" "
        "--hashalgs=\"1 2 3 4 5 6\" "
        "--authmethods=\"1 2 3 4 5 6 7 8 64221 64222 64223 64224 65001 65002 65003 65004 65005 65006 65007 65008 65009 65010\" "
        "--dhgroups=\"1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18\"")
    parser.add_argument(
        "--ikepath",
        type=str,
        help="The FULL ike-scan path if it is not in the PATH variable and/or the name changed.")
    parser.add_argument(
        "-c",
        "--clientids",
        type=str,
        help="A file (dictionary) with a client ID per line to enumerate valid client IDs in Aggressive Mode. "
        "This test is not launched by default.")

    args = parser.parse_args()
    Logger.set_verbose(args.loglevel)

    if args.target:
        targets.append(args.target)

    if args.input:
        try:
            f = open(args.input, "r")
            targets.extend(f.readlines())
            f.close()
        except:
            logger.error(
                "The input file specified ('%s') could not be opened." %
                args.input)

    if args.output:
        try:
            f = open(args.output, "w")
            f.close()
        except:
            logger.error(
                "The output file specified ('%s') could not be opened/created." %
                args.output)

    if not targets:
        logger.error("You need to specify a target or an input file (-i)")
        parser.parse_args(["-h"])
        sys.exit(1)

    if args.ikepath:
        self.ikescan_path = args.ikepath

    if args.encalgs:
        ENCLIST = args.encalgs.split()
        for alg in ENCLIST:
            parts = alg.split('/')
            for p in parts:
                if not p.isdigit():
                    logger.fatal(
                        "Wrong syntax for the encalgs parameter. Check syntax.")
                    parser.parse_args(["-h"])
                    sys.exit(1)

    if args.hashalgs:
        HASHLIST = args.hashalgs.split()
        for alg in HASHLIST:
            if not alg.isdigit():
                logger.fatal(
                    "Wrong syntax for the hashalgs parameter. Check syntax.")
                parser.parse_args(["-h"])
                sys.exit(1)

    if args.authmethods:
        AUTHLIST = args.authmethods.split()
        for alg in AUTHLIST:
            if not alg.isdigit():
                logger.fatal(
                    "Wrong syntax for the authmethods parameter. Check syntax.")
                parser.parse_args(["-h"])
                sys.exit(1)

    if args.dhgroups:
        GROUPLIST = args.dhgroups.split()
        for alg in GROUPLIST:
            if not alg.isdigit():
                logger.fatal(
                    "Wrong syntax for the dhgroups parameter. Check syntax.")
                parser.parse_args(["-h"])
                sys.exit(1)

    if args.xml:
        XMLOUTPUT = args.xml
    try:
        f = open(XMLOUTPUT, "w")
        f.close()
    except:
        logger.error("The XML output file could not be opened/created.")

    if args.clientids:
        try:
            f = open(args.clientids, "r")
            f.close()
            CLIENTIDS = args.clientids
        except:
            logger.error(
                "\033[91m[*]\033[0m The client ID dictionary could not be read. This test won't be launched.")

    if args.delay:
        DELAY = args.delay

    if args.fullalgs:
        ENCLIST = FULLENCLIST
        HASHLIST = FULLHASHLIST
        AUTHLIST = FULLAUTHLIST
        GROUPLIST = FULLGROUPLIST

    return args, targets


def main(argv=sys.argv[1:]):

    banner()
    vpns = {}
    args, targets = get_arguments()

    if os.geteuid():
        logger.fatal("got r00t?")
        return 10

    wrapper = ike_scan_wrapper.IkeScanWrapper(output=args.output)

    logger.info("Starting at %s" % time.strftime(
            "%a, %d %b %Y %H:%M:%S",time.localtime()))

    # 1. Discovery
    wrapper.discovery(args, targets, vpns)
    wrapper.check_ike_v2(args, targets, vpns)

    if not len(vpns.keys()):
        logger.critical("No IKE service was found")
        sys.exit(0)

    # 2. Fingerprint by checking VIDs and by analysing the service responses
    wrapper.fingerprint_VID(args, vpns)
    wrapper.fingerprint_show_backoff(args, vpns)
    # 3. Ciphers
    wrapper.check_encription_algorithms(args, vpns)
    # 4. Aggressive Mode
    wrapper.check_aggressives(args, vpns)
    # 5. Enumerate client IDs
    wrapper.enumerate_group_id(args, vpns)
    # . Parse the results
    wrapper.parse_output(args, vpns)

    logger.info(
        "Finished at %s" %
        time.strftime(
            "%a, %d %b %Y %H:%M:%S",
            time.localtime()))

    return 0

if __name__ == "__main__":
    sys.exit(main(sys.argv[1:]))
